### PR TITLE
Add static Reflect function to Vector3

### PR DESCRIPTION
**Pull request by AlvaroHerreroDev**
This adds a static function to Vector3 to Reflect a vector off the plane defined by a normal

### CODE REVIEW

File: packages/dev/core/src/Maths/math.vector.ts [hunk 1]
```diff
@@ -2515,6 +2515,21 @@ export class Vector3 {
         return result;
     }
 
+    /**
+     * Reflects a vector off the plane defined by a normal
+     * @param inDirection defines the vector direction
+     * @param normal defines the normal
```
Comment (Popov72) [inline packages/dev/core/src/Maths/math.vector.ts:2521]: You should indicate that the normal must be normalized for the result to be correct.

File: packages/dev/core/src/Maths/math.vector.ts [hunk 2]
```diff
@@ -2515,6 +2515,21 @@ export class Vector3 {
...
+     * @returns the resulting vector
+     */
+    public static Reflect<T extends Vector3>(inDirection: DeepImmutable<Vector3>, normal: DeepImmutable<Vector3>): T {
+        const _inDirection = new (inDirection.constructor as Vector3Constructor<T>)();
+        const _normal = new (normal.constructor as Vector3Constructor<T>)();
+        _inDirection.copyFrom(inDirection);
+        _normal.copyFrom(normal).normalize();
+
+        return _inDirection.subtract(_normal.scale(2 * Vector3.Dot(_inDirection, _normal)));
```
Comment (Popov72) [inline packages/dev/core/src/Maths/math.vector.ts:2530]: You should avoid creating too many temporaries to limit GC. Also, you should add a `toRef` variant, so something like:

```typescript
public static Reflect<T extends Vector3>(inDirection: DeepImmutable<Vector3>, normal: DeepImmutable<Vector3>): Vector3 {
    return this.ReflectToRef(inDirection, normal, new Vector3());
}

public static ReflectToRef<T extends Vector3>(inDirection: DeepImmutable<Vector3>, normal: DeepImmutable<Vector3>, ref: T): T {
    const tmp = TmpVectors.Vector3[0];
    tmp.copyFrom(normal).scaleInPlace(2 * Vector3.Dot(inDirection, normal));

    return ref.copyFrom(inDirection).subtractInPlace(tmp);
}
```

**Review by Popov72 — CHANGES_REQUESTED** [review]

File: packages/dev/core/src/Maths/math.vector.ts (hunk 1 above)
Comment (AlvaroHerreroDev) [reply packages/dev/core/src/Maths/math.vector.ts:2521]: Ok, but should I still normalize in the code?

Comment (sebavan) [reply packages/dev/core/src/Maths/math.vector.ts:2521]: Let s not normalize in the code and just add it in the code ?

**Review by sebavan — APPROVED** [review]
Only the little doc to add and good to go :-)

File: packages/dev/core/src/Maths/math.vector.ts (hunk 1 above)
Comment (Popov72) [reply packages/dev/core/src/Maths/math.vector.ts:2521]: You mean add it in the doc?

Comment (sebavan) [reply packages/dev/core/src/Maths/math.vector.ts:2521]: yup

**Comment by sebavan** [comment]
@AlvaroHerreroDev can you add the normalized comment to the doc of the normal parameter ?

File: packages/dev/core/src/Maths/math.vector.ts (hunk 1 above)
Comment (AlvaroHerreroDev) [reply packages/dev/core/src/Maths/math.vector.ts:2521]: Ok, removing the normalize in the actual code and updating the doc

**Review by sebavan — APPROVED** [review]

**Review by Popov72 — APPROVED** [review]

**Pull request merged**